### ISSUE TITLE
[Bug]: 03b halts when using family data

### ISSUE THREAD

**Issue by ks164** [genetics-of-dna-methylation-consortium/godmc_phase2#68]
### Contact Details

ks164@duke.edu

### Scripts

03b-methylation_adjustment1.sh

### What happened?

adjustment of methylation data at step 03b fails for my family data (MZ and DZ twins). 

I run the script in parallel, and each of the 100 batches are halted with one of the two following errors:

./03b-methylation_adjustment1.sh: line 78: 1987504 Killed                  ${R_directory}Rscript resources/methylation/adjust_pedigree.R ${home_directory} ${methylation_no_outliers} ${grmfile_relateds} ${covariates_combined}.txt ${transformed_methylation_adjusted} transformed ${nthreads} ${meth_chunks} ${i} ${methylation_array}

OR

Probe 1 of 91
Error in `.rowNamesDF<-`(x, value = value) : invalid 'row.names' length
Calls: main ... row.names<- -> row.names<-.data.frame -> .rowNamesDF<-
Execution halted

-----
I did notice that the eigenvectors calculated seem to be zero:

Reading methylation data...
Reading IDs
Reading GRM
Reading N
Creating data frame
Kinship matrix is 1502 by 1502
Calculating eigenvectors
list
List of 2
 $ values : num [1:1502] 4.32 4.31 4.29 4.27 4.16 ...
 $ vectors: num [1:1502, 1:1502] 0 0 0 0 0 0 0 0 0 0 ...
NULL

Attached are two example log files (the output is too big to paste)

[log.txt1.txt](https://github.com/user-attachments/files/17649450/log.txt1.txt)
[log.txt100.txt](https://github.com/user-attachments/files/17649451/log.txt100.txt)

### R version

4.3.3 (February, 2024)

### Python version

None

**Comment by ZXiaopu** [comment]
Hi,
The issue may raise from the low memory usage. Can you increase the memory and try again? (eg. in slurm system, increase the --mem)

**Comment by ks164** [comment]
Hi,

I have upped my memory usage, and every batched job now returns just the second error:

Error in `.rowNamesDF<-`(x, value = value) : invalid 'row.names' length
Calls: main ... row.names<- -> row.names<-.data.frame -> .rowNamesDF<-

**Comment by JennyvD** [comment]
Hi, I am having the exact same problem (also data from MZ and DZ twins).

**Comment by epzjlm** [comment]
Could I check whether you have chrX and ChrY probes?

**Comment by JennyvD** [comment]
Yes, confirmed (note that I checked processed_data/genetic_data/data.fam)

**Comment by drsora** [comment]
I have a similar issue with the one cohort I have with related data. Eigen vectors are 0 and the script freezes at probe 16

`Reading methylation data...
Reading IDs
Reading GRM
Reading N
Creating data frame
Kinship matrix is 1206 by 1206
Calculating eigenvectors
list
List of 2
 $ values : num [1:1206] 6.96 6.91 6.9 6.87 5.91 ...
 $ vectors: num [1:1206, 1:1206] 0 0 0 0 0 0 0 0 0 0 ...
NULL
Running with 16 threads
Probe 1 of 703622
Probe 2 of 703622
Probe 3 of 703622
Probe 4 of 703622
Probe 5 of 703622
Probe 6 of 703622
Probe 7 of 703622
Probe 8 of 703622
Probe 9 of 703622
Probe 10 of 703622
Probe 11 of 703622
Probe 12 of 703622
Probe 13 of 703622
Probe 14 of 703622
Probe 15 of 703622
Probe 16 of 703622
`

**Comment by epzjlm** [comment]
We found an error in 03b eg $home_directory should be scripts_directory. Does this solve your problem?

        ${R_directory}Rscript resources/methylation/adjust_pedigree.R \
                ${scripts_directory} \
                ${methylation_no_outliers} \
                ${grmfile_relateds} \
                ${covariates_combined}.txt \
                ${transformed_methylation_adjusted} \
                transformed \
                ${nthreads} \
                ${meth_chunks} \
                ${i} \
                ${methylation_array}
    
        echo "Generating untransformed residuals"
        ${R_directory}Rscript resources/methylation/adjust_pedigree.R \
                ${scripts_directory} \
                ${methylation_no_outliers} \
                ${grmfile_relateds} \
                ${covariates_combined}.txt \
                ${untransformed_methylation_adjusted} \
                untransformed \
                ${nthreads} \
                ${meth_chunks} \
                ${i} \
                ${methylation_array}

**Comment by ks164** [comment]
Hi,

I have re-run the the script updating "$home_directory" to "${scripts_directory}" and the error persists and is the same as my first report. The eigenvectors are still all zero too.

**Comment by epzjlm** [comment]
ok thanks for testing. We found another bug which is now fixed in 03b relatedness branch. Could you try?

**Comment by ks164** [comment]
I have it running now- will take a while, I'll let you know what happens when it is finished.

**Comment by drsora** [comment]
> ok thanks for testing. We found another bug which is now fixed in 03b relatedness branch. Could you try?

The 03b-relatedness branch freezes for me at the same place as the main, after probe 16.

**Comment by ZXiaopu** [comment]
> > ok thanks for testing. We found another bug which is now fixed in 03b relatedness branch. Could you try?
> 
> The 03b-relatedness branch freezes for me at the same place as the main, after probe 16.

Are you running in array jobs? - According to your previous log file, it looks like all 703622 probes run in the same job. The expectation is that 703622 probes will be divided into 100 jobs.

If not sure, can you provide your input of 03b-relatedness with below code?

```
source config
echo ${R_directory}Rscript resources/methylation/adjust_pedigree.R \
            ${scripts_directory} \
            ${methylation_no_outliers} \
            ${grmfile_relateds} \
            ${covariates_combined}.txt \
            ${transformed_methylation_adjusted} \
            transformed \
            ${nthreads} \
            ${meth_chunks} \
            ${i} \
            ${methylation_array}
```

**Comment by drsora** [comment]
> Are you running in array jobs?

it's also stuck on 16 when I run it in a batch

**Comment by JennyvD** [comment]
Do we know if this issue only comes up in datasets with monozygotic twins? Or does it also come up in datasets that contain e.g. 50% genetic relatedness. What type of familial relationships are in your dataset @drsora?

**Comment by ZXiaopu** [comment]
The previous issue occurs when the sex chromosome CpGs are adjusted. @JennyvD @drsora 
03b works like this
  1/ adjust autosomal CpGs for males and females
  2/ adjust chrX CpGs for females
  3/ adjust chrX CpGs for males
  4/ adjust chrY CpGs for males
  
The bug arose from step 2 because the kinship matrixes of female/male only were not generated. Instead, the script use the kinship matrix of all samples. During the developing stage, I tested the scripts on an all-female dataset, so I did not notice this issue. **-- I have fixed this issue in 03b-relatedness branch and @ks164 is running it.**

We don't need to worry about the 0s in the eigenvectors are expected. 02a (genotype processing scripts) assign the relatedness 0 if the GRM do not surpass the relatedness threshold (rel_cutoff="0.125" in config file)

**Comment by ZXiaopu** [comment]
> > Are you running in array jobs?
> 
> it's also stuck on 16 when I run it in a batch

Can you show me you input?
```
source config
echo ${R_directory}Rscript resources/methylation/adjust_pedigree.R \
            ${scripts_directory} \
            ${methylation_no_outliers} \
            ${grmfile_relateds} \
            ${covariates_combined}.txt \
            ${transformed_methylation_adjusted} \
            transformed \
            ${nthreads} \
            ${meth_chunks} \
            ${i} \
            ${methylation_array}
```

**Comment by drsora** [comment]
> Can you show me you input?

/usr/bin/Rscript resources/methylation/adjust_pedigree.R /home/drsora/godmc_phase2 /data/sonja/godmc/g2/processed_data/methylation_data/methylation_no_outliers.Robj /data/sonja/godmc/g2/processed_data/genetic_data/data_relateds /data/sonja/godmc/g2/processed_data/methylation_data/all_covariates.txt /data/sonja/godmc/g2/processed_data/methylation_data/transformed_methylation_adjusted transformed 16 100 epic:epic2

I am also trying to run it on the 03b-relatedness branch.

@JennyvD I don't have twins, just some siblings.

**Comment by ZXiaopu** [comment]
> > Can you show me you input?
> 
> /usr/bin/Rscript resources/methylation/adjust_pedigree.R /home/drsora/godmc_phase2 /data/sonja/godmc/g2/processed_data/methylation_data/methylation_no_outliers.Robj /data/sonja/godmc/g2/processed_data/genetic_data/data_relateds /data/sonja/godmc/g2/processed_data/methylation_data/all_covariates.txt /data/sonja/godmc/g2/processed_data/methylation_data/transformed_methylation_adjusted transformed 16 100 epic:epic2
> 
> I am also trying to run it on the 03b-relatedness branch.
> 
> [@JennyvD](https://github.com/JennyvD) I don't have twins, just some siblings.

Can you run 03b in unrelated samples well?
Can you send me the log file when you run 03b on related samples?

**Comment by drsora** [comment]
> Can you run 03b in unrelated samples well?
> Can you send me the log file when you run 03b on related samples?

Yes, I ran 3 just fine with my other 3 unrelated cohorts.

Here's the log:

[log1.txt](https://github.com/user-attachments/files/18655077/log1.txt)

**Comment by ZXiaopu** [comment]
> > Can you run 03b in unrelated samples well?
> > Can you send me the log file when you run 03b on related samples?
> 
> Yes, I ran 3 just fine with my other 3 unrelated cohorts.
> 
> Here's the log:
> 
> [log1.txt](https://github.com/user-attachments/files/18655077/log1.txt)

Can you try to increase the memory or nodes when you run this job? There is no error in the log file, so I wonder if it's because of the system setting.

**Comment by drsora** [comment]
> ncrease the memory or nodes

hmmmm I increased the nodes to 20, and now it freezes at 20 (previously, nodes were 16)

**Comment by ZXiaopu** [comment]
> In the end, 27 or the 100 batch jobs failed with the error message I posted in my last comment.
> 
> for the jobs that ran successfully, the number of expected files in the /processed_data/methylation_data/ folder has doubled; all have files with both .RData AND .RData_classes extensions

Thanks for testing. Files with .RData and .RData_classes are expected.

Do you have missing values in any of the input data for 03b? i.e. ${methylation_no_outliers}, ${grmfile_relateds}, ${covariates_combined}.txt

**Comment by ks164** [comment]
yes- it turns out there is ~17% missing data in methylation_no_outliers.Robj

I checked that there was no missing data in the original input data for 03a- is it expected that missing data would be introduced?

**Comment by ZXiaopu** [comment]
The issue might be in the regression model there are some factors with one unique values. Can you show me the output of the below code?

1/ Check whether these files exist. 
```
ls *.Female.chrX.*
ls *.Male.chrX.*
ls *.Male.chrY.*
```

2/ For any of the failed jobs, eg, the job id is 1, can you check the files
```
ls *.1.RData
```

3/ can you also run the attached script in the GoDMC scripts path?
```
mv check_cov.txt check_cov.R
source config
${R_directory}Rscript check_cov.R
```

[check_cov.txt](https://github.com/user-attachments/files/18732823/check_cov.txt)

**Comment by drsora** [comment]
> > ncrease the memory or nodes
> 
> hmmmm I increased the nodes to 20, and now it freezes at 20 (previously, nodes were 16)

@ZXiaopu  any ideas about this?

**Comment by ks164** [comment]
to answer your questions:

1/ Check whether these files exist.
These files exist at processed_data/methylation_data for all the jobs that ran successfully

ls *.Female.chrX.* N = 100 transformed, N = 73 untransformed RData files
ls *.Male.chrX.* N = 100 transformed, N = 73 untransformed RData files
ls *.Male.chrY.* N = 22 transformed, n = 22 untransformed RData files


2/ For any of the failed jobs, eg, the job id is 1, can you check the files 
This is the str of the files from one of my failed jobs- there are 3 transformed .RData and .RData_classes files but no untransformed files exist (N.B. I removed the IDs from the column descriptions)
/processed_data/methylation_data/transformed_methylation_adjusted.23.RData
 str(betas)
 num [1:4177, 1:1502] -5.72e-04 -1.18e-03 1.25e-03 -1.76e-03 -2.06e-05 ...
 - attr(*, "dimnames")=List of 2
  ..$ : chr [1:4177] "cg05346688" "cg05346732" "cg05346831" "cg05346832" ...
  ..$ : chr [1:1502] 

/processed_data/methylation_data/transformed_methylation_adjusted.Male.chrX.23.RData
 str(betas)
 num [1:127, 1:762] -0.374 0.112 -0.495 -0.892 0.363 ...
 - attr(*, "dimnames")=List of 2
  ..$ : chr [1:127] "cg05347431" "cg05356800" "cg05361475" "cg05363666" ...
  ..$ : chr [1:762] 

/processed_data/methylation_data/transformed_methylation_adjusted.Female.chrX.23.RData
 str(betas)
 num [1:127, 1:740] -0.816 0.379 0.52 -0.43 -0.171 ...
 - attr(*, "dimnames")=List of 2
  ..$ : chr [1:127] "cg05347431" "cg05356800" "cg05361475" "cg05363666" ...
  ..$ : chr [1:740] 

3/ can you also run the attached script in the GoDMC scripts path?
I couldn't get the script to run from command line, so I ran it within R on /processed_data/methylation_data/all_covariates.txt instead. The output is attached. N.B. we have MZ twins in our data, so the number of unique observations for e.g. genetic PCs won't necessarily be the same as the N individuals.

[all_covariates_counts.txt](https://github.com/user-attachments/files/18737986/all_covariates_counts.txt)

**Comment by ZXiaopu** [comment]
> > > ncrease the memory or nodes
> > 
> > 
> > hmmmm I increased the nodes to 20, and now it freezes at 20 (previously, nodes were 16)
> 
> [@ZXiaopu](https://github.com/ZXiaopu) any ideas about this?

Does it work well if you set nthreads to 1? - If so, maybe stick to nthreads=1, it will be slower but the running time is still acceptable.

**Comment by ks164** [comment]
Hi,

I ran the script with the failed job array I listed above.  It did seem to output lm results successfully with some warnings 

[Check03b_output.txt](https://github.com/user-attachments/files/18770180/Check03b_output.txt)

I then ran the code in part 2 on the same failed job array, and it resulted in the following untransformed data files (it did take 18+ hours to run the job):

 load("./processed_data/methylation_data/untransformed_methylation_adjusted.23.RData")
 str(betas)
 num [1:4177, 1:1502] -3.99e-06 -6.57e-06 1.83e-05 -1.12e-05 -3.72e-07 ...
 - attr(*, "dimnames")=List of 2
  ..$ : chr [1:4177] "cg05346688" "cg05346732" "cg05346831" "cg05346832" ...
  ..$ : chr [1:1502] 

 load("./processed_data/methylation_data/untransformed_methylation_adjusted.Male.chrX.23.RData")
 str(betas)
 num [1:127, 1:762] -0.00629 0.00456 -0.01044 -0.00844 0.00759 ...
 - attr(*, "dimnames")=List of 2
  ..$ : chr [1:127] "cg05347431" "cg05356800" "cg05361475" "cg05363666" ...
  ..$ : chr [1:762] 

 load("./processed_data/methylation_data/untransformed_methylation_adjusted.Female.chrX.23.RData")
 str(betas)
 num [1:127, 1:740] -0.03355 0.00951 0.01351 -0.01729 -0.00299 ...
 - attr(*, "dimnames")=List of 2
  ..$ : chr [1:127] "cg05347431" "cg05356800" "cg05361475" "cg05363666" ...
  ..$ : chr [1:740]

**Comment by ZXiaopu** [comment]
> Hi,
> 
> I ran the script with the failed job array I listed above. It did seem to output lm results successfully with some warnings
> 
> [Check03b_output.txt](https://github.com/user-attachments/files/18770180/Check03b_output.txt)
> 
> I then ran the code in part 2 on the same failed job array, and it resulted in the following untransformed data files (it did take 18+ hours to run the job):
> 
> load("./processed_data/methylation_data/untransformed_methylation_adjusted.23.RData") str(betas) num [1:4177, 1:1502] -3.99e-06 -6.57e-06 1.83e-05 -1.12e-05 -3.72e-07 ...
> 
> * attr(*, "dimnames")=List of 2
>   ..$ : chr [1:4177] "cg05346688" "cg05346732" "cg05346831" "cg05346832" ...
>   ..$ : chr [1:1502]
> 
> load("./processed_data/methylation_data/untransformed_methylation_adjusted.Male.chrX.23.RData") str(betas) num [1:127, 1:762] -0.00629 0.00456 -0.01044 -0.00844 0.00759 ...
> 
> * attr(*, "dimnames")=List of 2
>   ..$ : chr [1:127] "cg05347431" "cg05356800" "cg05361475" "cg05363666" ...
>   ..$ : chr [1:762]
> 
> load("./processed_data/methylation_data/untransformed_methylation_adjusted.Female.chrX.23.RData") str(betas) num [1:127, 1:740] -0.03355 0.00951 0.01351 -0.01729 -0.00299 ...
> 
> * attr(*, "dimnames")=List of 2
>   ..$ : chr [1:127] "cg05347431" "cg05356800" "cg05361475" "cg05363666" ...
>   ..$ : chr [1:740]

Great. It looks like rerunning the failed job can solve this issue. It does take a longer time to run relatedness adjustment, especially with ~1500 samples.

Can you run the part2 for the failed jobs as well please?

**Comment by ks164** [comment]
Sorry, do you mean run part 2 on the other 22 failed jobs as well?

**Comment by ZXiaopu** [comment]
Yes. If it worked for chunk 23, it is supposed to work for the remaining 22 jobs.

**Comment by ks164** [comment]
I have run the code on the 27 job arrays that didn't run first time, and they all produced untransformed RData files with the same structure (as above). However, the log files give this error:

Probe 1 of 762
Error in B[i, ] <- out$x :
  number of items to replace is not a multiple of replacement length
Calls: main -> run.adjust.cov -> adjust.relatedness.serial
In addition: There were 50 or more warnings (use warnings() to see the first 50)
Execution halted

**Comment by ZXiaopu** [comment]
Did this error occur in all 27 jobs or just one?

**Comment by ks164** [comment]
all of them- the probe count seems off in the error message - 762 is the number of Males in my dataset, not the number of probes per job array (4177 or 127)

**Comment by ZXiaopu** [comment]
> I have run the code on the 27 job arrays that didn't run first time, and they all produced untransformed RData files with the same structure (as above). However, the log files give this error:
> 
> Probe 1 of 762 Error in B[i, ] <- out$x : number of items to replace is not a multiple of replacement length Calls: main -> run.adjust.cov -> adjust.relatedness.serial In addition: There were 50 or more warnings (use warnings() to see the first 50) Execution halted

Did you get the error for all 27 jobs or specific jobs?
It looks like the error occurs when adjusting sex chromosome - Are you saying RData was produced for autosomal but not sex chr? i.e. untransformed_methylation_adjusted.[0-9].RData is available but untransformed_methylation_adjusted.*chrX.[0-9].RData were not?

**Comment by ks164** [comment]
I have 3 .RData files for all 27 previously failed jobs, including .*chrX.[0-9]. The structure of the three files is the same as that in my comment yesterday

**Comment by ZXiaopu** [comment]
> I have 3 .RData files for all 27 previously failed jobs, including .*chrX.[0-9]. The structure of the three files is the same as that in my comment yesterday

The issue maybe there is only 1 Y probe in the failed jobs, so the matrix dimension is not as expected. I updated the scripts to solve this situation. 

Can you check whether the chrY probe was adjusted for the transformed data? i.e. if job 1 failed and untransformed_methylation_adjusted.Male.chrY.1.RData was not generated, transformed_methylation_adjusted.Male.chrY.1.RData was not supposed to exist as well.

If this is the case, we can confirm where the bug is from. And then please run updated 03b (it will skip autosomal and chrX probes if these data have been successfully generated)

**Comment by ks164** [comment]
There are matched transformed/untransformed chrY RData files (i.e. they are either both missing or both generated as you indicated).

I will re-run the updated 03b script.

**Comment by drsora** [comment]
> > > > ncrease the memory or nodes
> > > 
> > > 
> > > hmmmm I increased the nodes to 20, and now it freezes at 20 (previously, nodes were 16)
> > 
> > 
> > [@ZXiaopu](https://github.com/ZXiaopu) any ideas about this?
> 
> Does it work well if you set nthreads to 1? - If so, maybe stick to nthreads=1, it will be slower but the running time is still acceptable.

It works with nthreads=1 but it took 3 days when it was 16 threads, I don't think the running time is still acceptable with 1, I can't hog our server for that long. It takes more than an hour to just run 1/100

**Comment by ZXiaopu** [comment]
> > > > > ncrease the memory or nodes
> > > > 
> > > > 
> > > > hmmmm I increased the nodes to 20, and now it freezes at 20 (previously, nodes were 16)
> > > 
> > > 
> > > [@ZXiaopu](https://github.com/ZXiaopu) any ideas about this?
> > 
> > 
> > Does it work well if you set nthreads to 1? - If so, maybe stick to nthreads=1, it will be slower but the running time is still acceptable.
> 
> It works with nthreads=1 but it took 3 days when it was 16 threads, I don't think the running time is still acceptable with 1, I can't hog our server for that long. It takes more than an hour to just run 1/100

How many samples do you have? 
Does it work if you change to other nodes? I just wonder if the frozen is due to the node.

**Comment by drsora** [comment]
@ZXiaopu Yes, batch 2 also fails. If I use 16 nodes, it fails on 16th probe. It's just so weird because I only have this problem in this cohort (only one with related people).

**Comment by ZXiaopu** [comment]
> [@ZXiaopu](https://github.com/ZXiaopu) Yes, batch 2 also fails. If I use 16 nodes, it fails on 16th probe. It's just so weird because I only have this problem in this cohort (only one with related people).

How many samples do you have? It's just odd that using 16 threads still takes 3 days. 
One more thing I would like to try is to change the partition. eg. some partition may be down to make it unable to run in parallel?
Can you also ask the IT team for some advice at the same time?

**Comment by ZXiaopu** [comment]
> > [@ZXiaopu](https://github.com/ZXiaopu) Yes, batch 2 also fails. If I use 16 nodes, it fails on 16th probe. It's just so weird because I only have this problem in this cohort (only one with related people).
> 
> @drsora How many samples do you have? It's just odd that using 16 threads still takes 3 days. One more thing I would like to try is to change the partition. eg. some partition may be down to make it unable to run in parallel? Can you also ask the IT team for some advice at the same time?

**Comment by drsora** [comment]
@ZXiaopu Around 2000 people. I don't think I can ask my IT team anything, they're historically very useless. I'm not running them on a closer because I couldn't get that to work, just normally like  ./03b-methylation_adjustment1.sh 1
That takes about 17 minutes per chunk with the other cohorts

**Comment by ZXiaopu** [comment]
> [@ZXiaopu](https://github.com/ZXiaopu) Around 2000 people. I don't think I can ask my IT team anything, they're historically very useless. I'm not running them on a closer because I couldn't get that to work, just normally like ./03b-methylation_adjustment1.sh 1 That takes about 17 minutes per chunk with the other cohorts

@drsora without running on a cluster might make it slow and fail to run in parallel, as adjusting the pedigree requires more memory to run than independent samples, and it is expected to be slower than independent samples.

I would suggest you to estimate how long your server takes to run with threads =1, and increase meth_chunks with threads = 1. This might be the easiest way for us to solve this issue.

**Comment by ks164** [comment]
Hi,

I managed to run the updated 03b script successfully and got 100 transformed/untransformed .RData and RData classes files as expected, except for Male.chrY where there are 11.

However, when I run aggregate_adjustment1.sh, only transformed/untransformed_methylation_adjusted.RData and transformed/untransformed_methylation_adjusted.Female.chrX.RData are created.

Aggregation on the male samples fails as the the number of CpGs before and after adjustment are NOT consistent:

"The number of CpGs before and after adjustment are inconsistent
Warning message:
In rownames(beta.sub) == rownames(alldat) :
  longer object length is not a multiple of shorter object length"

[.txt__aggregation.txt](https://github.com/user-attachments/files/18968279/default.txt__aggregation.txt)
[log.txt_Female.chrX_aggregation.txt](https://github.com/user-attachments/files/18968278/log.txt_Female.chrX_aggregation.txt)
[log.txt_Male.chrX_aggregation.txt](https://github.com/user-attachments/files/18968277/log.txt_Male.chrX_aggregation.txt)
[log.txt_Male.chrY_aggregation.txt](https://github.com/user-attachments/files/18968280/log.txt_Male.chrY_aggregation.txt)

**Comment by ZXiaopu** [comment]
@ks164 Hi, can you run the attached scripts? If you put them at the godmc scripts path, run bash aggregate_adjustment_temp.sh  Male.chrX and show me the output please?

[aggregate_adjustment1_temp.sh.txt](https://github.com/user-attachments/files/18993763/aggregate_adjustment1_temp.sh.txt)
[aggregate_chunks_temp.R.txt](https://github.com/user-attachments/files/18993762/aggregate_chunks_temp.R.txt)

**Comment by ks164** [comment]
this was the output:

Aggregating methylation_adjustment1 chunks
There were 17 warnings (use warnings() to see them)
Checking all files are present
Error in 1:n_chunks : NA/NaN argument
Calls: main -> paste0
Execution halted

**Comment by ZXiaopu** [comment]
> this was the output:
> 
> Aggregating methylation_adjustment1 chunks There were 17 warnings (use warnings() to see them) Checking all files are present Error in 1:n_chunks : NA/NaN argument Calls: main -> paste0 Execution halted

It looks like the config file was not used. Are the config file at the same path?

```
source config

**Comment by ks164** [comment]
I get the same error when I explicitly source the config as above.
They are not on the same path, but I still get the same error when I edit the location in the .sh file or run the scripts from the same path.

**Comment by ZXiaopu** [comment]
The scripts I sent earlier were to check the output of aggregated results. Maybe it will be easier if you add the code below to resources/methylation/aggregate_chunks.R line 71 to check the difference beween beta.sub and alldat.

```
print(dim(beta.sub))
print(dim(alldat))
print(rownames(beta.sub)[!rownames(beta.sub) %in% rownames(alldat)])
print(rownames(alldat)[!rownames(alldat) %in% rownames(beta.sub)])
``` 

They won't change

**Comment by ks164** [comment]
Here's the output from from running resources/methylation/aggregate_adjustment1.sh Male.chrX

[Aggregating methylation_adjustment1_output.txt](https://github.com/user-attachments/files/19056507/Aggregating.methylation_adjustment1_output.txt)

**Comment by epzjlm** [comment]
@drsora 

To me it looks like it goes wrong because the bash variables are not specified correctly. Could you print out the following:

echo ${R_directory}Rscript resources/methylation/adjust_pedigree.R \
		${scripts_directory} \
		${methylation_no_outliers} \
		${grmfile_relateds} \
		${covariates_combined}.txt \
		${transformed_methylation_adjusted} \
		transformed \
		${nthreads} \
		${meth_chunks} \
		${i} \
        	${methylation_array}

**Comment by drsora** [comment]
@epzjlm 

source config
echo ${R_directory}Rscript resources/methylation/adjust_pedigree.R
/usr/bin/Rscript resources/methylation/adjust_pedigree.R
echo ${scripts_directory}
/home/drsora/godmc_phase2
echo ${methylation_no_outliers}
/data/sonja/godmc/g2/processed_data/methylation_data/methylation_no_outliers.Robj
echo ${grmfile_relateds}
/data/sonja/godmc/g2/processed_data/genetic_data/data_relateds
echo ${covariates_combined}.txt
/data/sonja/godmc/g2/processed_data/methylation_data/all_covariates.txt
echo ${transformed_methylation_adjusted}
/data/sonja/godmc/g2/processed_data/methylation_data/transformed_methylation_adjusted
echo ${nthreads}
16
echo ${meth_chunks}
100
echo ${i}

echo ${methylation_array}
epic:epic2

**Comment by ZXiaopu** [comment]
> Here's the output from from running resources/methylation/aggregate_adjustment1.sh Male.chrX
> 
> [Aggregating methylation_adjustment1_output.txt](https://github.com/user-attachments/files/19056507/Aggregating.methylation_adjustment1_output.txt)

We spotted a error related to this. Can you delete all male CpG adjustment RData and rerun the 03b in branch male_sex_chr_cpg?

```
cd godmc
rm processed_data/methylation_data/*.Male.*
sbatch 03b [1-100]
```

Given you have the autosomal CpG and Female chrX results, 03b will skip this steps and only run male.chrX and male.chrY

**Comment by epzjlm** [comment]
@ks164 @ZXiaopu Is there any update on this?

**Comment by Ohanna-uoft** [comment]
Hi. I have a similar problem in 03b. The transformed/untransformed .RData are ok for autosomal and Male.chrY, but not for Male.chrX and Female.chrX: "The number of CpGs before and after adjustment are not consistent".
Can you point me to known fixes?

**Comment by ks164** [comment]
I have finally managed to complete 03 BUT I now have 2 further issues:

1. lambda is very high for the positive control ~1.5- is this due to MZ/DZ twin data structure?
2. I am unable to upload the results as the qqplots have been saved as jpegs and not pngs, meaning the upload check fails

**Comment by JennyvD** [comment]
Hi Karen, I also have MZ/DZ twin data, but no high lambda for the positive control

**Comment by ks164** [comment]
Thanks Jenny! I think there's something more than the twin structure causing such a high lambda then.
